lib/common/client/middleware/request/immutable_headers.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+module Common
+  module Client
+    module Middleware
+      module Request
+        class ImmutableHeaders < Faraday::Middleware
+          HTTP_PROTOCOL_HEADERS = %w(
+            Accept
+            Accept-Charset
+            Accept-Encoding
+            Accept-Language
+            Accept-Ranges
+            Age
+            Allow
+            Authorization
+            Cache-Control
+            Connection
+            Content-Encoding
+            Content-Language
+            Content-Length
+            Content-Location
+            Content-MD5
+            Content-Range
+            Content-Type
+            Date
+            ETag
+            Expect
+            Expires
+            From
+            Host
+            If-Match
+            If-Modified-Since
+            If-None-Match
+            If-Range
+            If-Unmodified-Since
+            Last-Modified
+            Location
+            Max-Forwards
+            Pragma
+            Proxy-Authenticate
+            Proxy-Authorization
+            Range
+            Referer
+            Retry-After
+            Server
+            TE
+            Trailer
+            Transfer-Encoding
+            Upgrade
+            User-Agent
+            Vary
+            Via
+          ).freeze
+
+          def call(env)
+            headers = {}
+            env.request_headers.each do |k, v|
+              if HTTP_PROTOCOL_HEADERS.include?(k)
```

Comment:
Minor, but lookup with include? is O(n) vs O(1) with a Set

Reply:
just ignore this - optimization here is seriously negligible 😉

Reply:
There's ~10'ish headers per request and this gets called on every request.  We might be talking picoseconds for all I know :laughing: , but it would give me a warm fuzzy to use a Set instead.

Reply:
Also nitpicky, but using `.map!` instead of `.each` could save a few LoC:
```ruby
env.request_headers.map! do |k, v|
  if HTTP_PROTOCOL_HEADERS.exclude?(k)
    headers[ImmutableString.new(k)] = v
  end
end
```